src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalysisScope.cs

```diff
@@ -85,6 +86,22 @@ private AnalysisScope(IEnumerable<SyntaxTree> trees, IEnumerable<AdditionalText>
             Debug.Assert(isPartialAnalysis || FilterSpanOpt == null);
             Debug.Assert(isPartialAnalysis || !isSyntacticSingleFileAnalysis);
 
+            if (filterSpanOpt.HasValue)
+            {
+                Debug.Assert(filterFile.HasValue);
+                Debug.Assert(filterFile.Value.SourceTree != null);
+                Debug.Assert(filterSpanOpt.Value.Length <= filterFile.Value.SourceTree.Length);
+
+                // PERF: Clear out filter span if the span length is equal to the entire tree span.
+                //       We are basically analyzing the entire tree, and clearing out the filter span
+                //       avoids span intersection checks for each symbol/node/operation in the tree
+                //       to determine if it falls in the analysis scope.
+                if (filterSpanOpt.Value.Length == filterFile.Value.SourceTree.Length)
```

Comment:
>if (filterSpanOpt.Value.Length == filterFile.Value.SourceTree.Length) [](http://example.com/codeflow?start=16&length=69)

What makes us think that ```filterSpanOpt.Value.Start``` is 0?

Reply:
Ah, good point. Given that this is coming from a public API, we cannot make this assumption. I'll add the check for SpanStart to be `0`.